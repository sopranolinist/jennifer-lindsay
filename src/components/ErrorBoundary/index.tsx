import {
  createStyles,
  Theme,
  withStyles,
  WithStyles
} from '@material-ui/core/styles';
import Typography from '@material-ui/core/Typography';
import React, { Component, ReactChild } from 'react';
import DocumentTitle from 'react-document-title';
import { getPageTitle } from '../../helpers/DataHelper';

const styles = (theme: Theme) =>
  createStyles({
    root: {
      alignItems: 'baseline',
      padding: theme.spacing(3),
      paddingTop: theme.spacing(12),
      flexGrow: 1,
      flexDirection: 'column',
      display: 'flex',
      textAlign: 'left'
    },
    errorHeader: {
      color: theme.palette.error['main'],
      marginBottom: theme.spacing(2)
    },
    errorMessage: {
      marginBottom: theme.spacing(2)
    }
  });

interface Props extends WithStyles<typeof styles> {
  children?: ReactChild;
  hasError?: string;
}
interface State {
  error: boolean;
}

class ErrorBoundary extends Component<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = { error: false };
  }

  static getDerivedStateFromError(error: Error | null) {
    return { error: true };
  }

  componentDidCatch(error: Error | null, info: object) {
    // console.log('ErrorBoundary.componentDidCatch', error, info);
  }

  render() {
    const pageTitle = 'Error';
    const { error } = this.state;
    const { children, hasError, classes } = this.props;
    if (!error && !hasError) return children;
    return (
      <div className={classes.root}>
        <DocumentTitle title={getPageTitle(pageTitle)} />
        <Typography component="h1" variant="h6" className={classes.errorHeader}>
          Error
        </Typography>
        <Typography
          component="span"
          variant="body1"
          className={classes.errorMessage}
        >
          The site has encountered an error{hasError && ':'}
        </Typography>
        {hasError && (
          <Typography
            component="span"
            variant="body1"
            className={classes.errorMessage}
          >
            {hasError}
          </Typography>
        )}
      </div>
    );
  }
}

export default withStyles(styles)(ErrorBoundary);
