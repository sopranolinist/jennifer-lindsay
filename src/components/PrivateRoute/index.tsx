import React from 'react';
import { Route, RouteProps } from 'react-router-dom';
import Unauthorized from '../../screens/Unauthorized';

interface Props extends RouteProps {}

// eslint-disable-next-line import/no-anonymous-default-export
export default ({ component, ...rest }: Props) => {
const isAuthorized = true;
  return (
    <Route
      {...rest}
      component={isAuthorized ? component : () => <Unauthorized />}
    />
  );
};
