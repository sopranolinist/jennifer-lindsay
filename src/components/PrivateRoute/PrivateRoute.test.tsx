import { render } from '@testing-library/react';
import React from 'react';
import {
  authExamples,
  getFullyWrappedComponent,
  getJssWrappedComponent,
  getRouterWrappedComponent,
  mockConsole
} from '../../helpers/TestHelper';
import ErrorBoundary from '../ErrorBoundary';
import PrivateRoute from './';

const HelloWorld = () => <span>Hello World!</span>;

describe('Component - PrivateRoute', () => {
  beforeEach(() => {
    mockConsole();
  });

  test('fails outside of Router', async done => {
    const { container } = render(
      getJssWrappedComponent(
        <ErrorBoundary>
          <PrivateRoute exact path={'/'} component={HelloWorld} />
        </ErrorBoundary>
      )
    );
    expect(container).toMatchSnapshot();
    done();
  });

  test('should enforce auth context', async done => {
    const { container } = render(
      getJssWrappedComponent(
        getRouterWrappedComponent(
          <ErrorBoundary>
            <PrivateRoute exact path={'/'} component={HelloWorld} />
          </ErrorBoundary>
        )
      )
    );
    expect(container).toMatchSnapshot();
    done();
  });

  test('renders route for authorized user', async done => {
    const { authContextProps, authObjProps } = authExamples.signedInAuthorized;
    const { container } = render(
      getFullyWrappedComponent(
        <PrivateRoute exact={true} path={'/'} component={HelloWorld} />,
        { authContextProps, authObjProps }
      )
    );
    expect(container).toMatchSnapshot();
    done();
  });

  test('renders unauthorized screen for unauthorized user', async done => {
    const {
      authContextProps,
      authObjProps
    } = authExamples.signedInUnauthorized;
    const { container } = render(
      getFullyWrappedComponent(
        <PrivateRoute exact={true} path={'/'} component={HelloWorld} />,
        { authContextProps, authObjProps }
      )
    );
    expect(container).toMatchSnapshot();
    done();
  });
});
