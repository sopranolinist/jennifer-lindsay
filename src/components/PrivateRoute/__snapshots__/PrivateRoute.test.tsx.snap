// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component - PrivateRoute fails outside of Router 1`] = `
<div>
  <div
    class="ErrorBoundary-root-1"
  >
    <h1
      class="MuiTypography-root ErrorBoundary-errorHeader-2 MuiTypography-h6"
    >
      Error
    </h1>
    <span
      class="MuiTypography-root ErrorBoundary-errorMessage-3 MuiTypography-body1"
    >
      The site has encountered an error
    </span>
  </div>
</div>
`;

exports[`Component - PrivateRoute renders route for authorized user 1`] = `
<div>
  <span>
    Hello World!
  </span>
</div>
`;

exports[`Component - PrivateRoute renders unauthorized screen for unauthorized user 1`] = `
<div>
  <span>
    Hello World!
  </span>
</div>
`;

exports[`Component - PrivateRoute should enforce auth context 1`] = `
<div>
  <div
    class="ErrorBoundary-root-1"
  >
    <h1
      class="MuiTypography-root ErrorBoundary-errorHeader-2 MuiTypography-h6"
    >
      Error
    </h1>
    <span
      class="MuiTypography-root ErrorBoundary-errorMessage-3 MuiTypography-body1"
    >
      The site has encountered an error
    </span>
  </div>
</div>
`;
